stdlib/public/TensorFlow/Dataset.swift

```diff
@@ -0,0 +1,141 @@
+//===-- Dataset.swift -----------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// The dataset API.
+//
+//===----------------------------------------------------------------------===//
+
+/// Represents a potentially large set of elements.
+///
+/// A `SingleValueDataset` can be used to represent an input pipeline as a
+/// collection of element tensors.
+@_fixed_layout
+public struct SingleValueDataset {
+  // NOTE: SingleValueDataset is currently specialzed on `Float` to work around
+  // deabstraction bug SR-8463. This is to avoid indirect passing in SIL. When
+  // the bug gets resolved, `SingleValueDataset` will have generic parameter
+  // `Scalar`.
+  public typealias Scalar = Float
+  @usableFromInline let _handle: VariantHandle
+  public let elementShape: TensorShape
+
+  @usableFromInline @inline(__always)
+  internal init(_handle: VariantHandle, elementShape: TensorShape) {
+    self._handle = _handle
+    self.elementShape = elementShape
+  }
+}
+
+public extension SingleValueDataset {
+  /// Creates a dataset from a batch of elements as a tensor.
+  ///
+  /// - Parameter
+  ///   - elements: The batch of elements.
+  ///   - elementShape: The shape that the 
+  ///
+  // FIXME: Due to inability to perform shape inference during deabstraction,
+  // users must specify the element shape for today.
```

Comment:
This isn't deabstraction's fault - this is XLA not supporting dynamic shapes.  Deabstraction can't magically know the shape of the dataset you've constructed with this API.

Reply:
I'm aware that deabstraction cannot handle shapes today. Maybe I did not use the best word to describe this. TensorFlow has shape inference C APIs, and you can call that to get the shape of some graph node, if not specified by user code. This idea was brought up by Anthony a few days ago - maybe it's worth considering delaying constant checking of shape attributes to graph lowering so that TF APIs can help.